src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -521,48 +512,52 @@ public int hashCode() {
                 unsafeGetBase());
     }
 
-    public static AbstractMemorySegmentImpl ofBuffer(Buffer bb) {
-        Objects.requireNonNull(bb);
-        Object base = NIO_ACCESS.getBufferBase(bb);
-        if (!bb.isDirect() && base == null) {
+    @ForceInline
+    public static AbstractMemorySegmentImpl ofBuffer(Buffer b) {
+        final int scaleFactor = NIO_ACCESS.scaleFactor(b);
+        return ofBuffer(b, offset(b, scaleFactor), length(b, scaleFactor));
+    }
+
+    @ForceInline
+    private static AbstractMemorySegmentImpl ofBuffer(Buffer b, long offset, long length) {
+        final Object base = NIO_ACCESS.getBufferBase(b);
+        return (base == null)
+                ? arrayFreeSegment(b, offset, length)
+                : NIO_ACCESS.arrayBackedSegment(b, base, offset, length, b.isReadOnly(), bufferScope(b));
+    }
+
+    @ForceInline
+    private static long offset(Buffer b, int scaleFactor) {
+        final long bbAddress = NIO_ACCESS.getBufferAddress(b);
+        return bbAddress + (((long) b.position()) << scaleFactor);
+    }
+
+    @ForceInline
+    private static long length(Buffer b, int scaleFactor) {
+        return ((long) b.limit() - b.position()) << scaleFactor;
+    }
+
+    @ForceInline
+    private static AbstractMemorySegmentImpl arrayFreeSegment(Buffer b, long offset, long length) {
```

Comment:
the names `arrayFreeSegment` and `arrayBackedSegments` seem a bit confusing. I'd suggest `ofDirectBuffer` and `ofHeapBuffer`

Reply:
I called them that initially but there is a subtle difference; there are heap buffers that are not backed by an array. The name arrayFreeSegment could better be arrayLessSegment.

Reply:
Not sure I follow. The code asks if there's a backing array (accessing base). Then:
* if there is a base, it calls `arrayBackedSegment`
* otherwise it calls `arrayLessSegment`.

Now, `arrayBackedSegment` is only implemented for heap buffers. Direct buffers throws UOE.
And, inside `arrayLessSegment` we throw if we see a non-direct (=heap) buffer.

So... it seems to me that arrayBacked = heap and arrayLess = direct?

Reply:
> Do you refer to char buffers backed by Strings?

If so, I wouldn't worry much about that -- after all, the logic in the code already detects this condition and throws an exception... so it's still heap buffer vs. direct buffer IMHO